File: src/main/Java/dfs/boj11725.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;
import java.util.TreeMap;
import java.util.stream.IntStream;

public class boj11725 {private static ArrayList<Integer>[] linked;
    private static boolean[] isVisited;
    private static TreeMap<Integer, Integer> result = new TreeMap<>();
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(reader.readLine());
        linked = new ArrayList[n + 1];
        IntStream.range(1, n + 1).forEach(index -> linked[index] = new ArrayList<>());
        isVisited = new boolean[n + 1];
        result.put(1, 0);
        for (int i = 1; i < n; i++) {
            StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
            int x = Integer.parseInt(tokenizer.nextToken());
            int y = Integer.parseInt(tokenizer.nextToken());
            linked[x].add(y);
            linked[y].add(x);
        }
        StringBuilder builder = new StringBuilder();
        dfs(1);
        result.remove(1);
        result.forEach((key, value) -> builder.append(value).append('\n'));
        writer.append(builder).close();
    }
    private static void dfs(int n) {
        if (isVisited[n]) return;
        isVisited[n] = true;
        linked[n].forEach(x -> {
            if (result.containsKey(x)) result.put(n, x);
            else result.put(x, n);
            dfs(x);
        });
    }
}

File: src/main/kotlin/dfs/boj11725.kt

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

private var linked = arrayOf(ArrayList<Int>())
private var isVisited = booleanArrayOf()
private val result = TreeMap<Int, Int>();
fun main() {
    val r = BufferedReader(InputStreamReader(System.`in`))
    val w = BufferedWriter(OutputStreamWriter(System.out))
    val n = r.readLine().toInt()
    linked = Array(n + 1) { ArrayList() }
    isVisited = BooleanArray(n + 1);
    repeat(n - 1) {
        val (x, y) = r.readLine().split(" ").map { it.toInt() }
        linked[x].add(y)
        linked[y].add(x)
    }
    result[1] = 0
    dfs(1)
    result.remove(1)
    result.forEach{ key, value -> w.appendLine(value.toString()) }
    w.close()
}

private fun dfs(n : Int) {
    if (isVisited[n]) return
    isVisited[n] = true
    linked[n].forEach{
        if (result.containsKey(it)) result.put(n, it)
        else result.put(it, n)
        dfs(it)
    }
}